### PR TITLE
Member roles

### PR DESCRIPTION
**Pull request by FalkoJoseph**
Implemented member roles.

screen shot 2015-09-08 at 01 07 38

screen shot 2015-09-08 at 01 08 41

Features

* 2 member roles: owner (admin) and contributor.
* You can't add yourself to an organisation, only admins can.
* Admins can't remove themselves.
* Contributors can't add others to a member group.
* Only admins can edit/close/delete an organisation and event.

Also created an application helper to quickly access the member role from the current user on a specific organisation.

     organisation_owner(@organisation)

Still needs the new parameter implementation and some cleanup though. Other than that, the feature itself is ready.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -18,6 +18,7 @@ gem "sass-rails", "~> 5.0"
 gem "stripe-rails"
 gem "turbolinks"
 gem "uglifier", ">= 1.3.0"
+gem "rolify"
```
Comment (JensDebergh) [inline Gemfile]: Sort by name please :8ball:

File: app/controllers/events_controller.rb [hunk 2]
```diff
@@ -26,6 +26,7 @@ def edit
 
   def create
     @event = Event.new(event_params)
+    @event.status = "open"
```
Comment (JensDebergh) [inline app/controllers/events_controller.rb]: Enter a default inside the model.

before_save :method_fill_in

File: app/controllers/organisations_controller.rb [hunk 3]
```diff
@@ -61,7 +66,6 @@ def set_organisation
   def organisation_params
     params.require(:organisation).
       permit(
-        :name, :description, :website, :picture, :locale, :organisation_users,
-        user_ids: [])
+        :name, :description, :website, :picture, :locale)
```
Comment (JensDebergh) [inline app/controllers/organisations_controller.rb]: Put this on the same line as permit. 80 characters max

File: app/controllers/registrations_controller.rb [hunk 4]
```diff
@@ -5,6 +5,7 @@ def create
     @organisation.users << resource
     @organisation.name = resource.email
     @organisation.save
+    resource.add_role :owner, @organisation
```
Comment (JensDebergh) [inline app/controllers/registrations_controller.rb]: Looks dirty ;D

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -1,7 +1,61 @@
...
 
   def show
+  end
+
+  def new
+    @user = User.new
+  end
+
+  def create
+    user = User.where(email: params[:email]).first
```
Comment (JensDebergh) [inline app/controllers/users_controller.rb]: Use find_by to get only 1 result.

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -1,7 +1,61 @@
...
+
+    if !@organisation.users.where(email: params[:email]).exists?
+      User.assign_role(user, params[:user_role], @organisation)
+      redirect_to organisation_users_url, notice:
+        "Member was succesfully added."
```
Comment (JensDebergh) [inline app/controllers/users_controller.rb]: Add translations using t(".relativetranslation")

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -1,23 +1,23 @@
 module ApplicationHelper
-  def bootstrap_class_for flash_type
+  def bootstrap_class_for(flash_type)
```
Comment (JensDebergh) [inline app/helpers/application_helper.rb]: Put an alert class inside the view and call the flash type like so 

<p class="alert-#{key}"></p> 

When interating the flash object we will get a key called succes, danger, notice.

File: app/helpers/application_helper.rb [hunk 8]
```diff
@@ -35,12 +35,20 @@ def display_name(id)
   def get_id
```
Comment (JensDebergh) [inline app/helpers/application_helper.rb]: I'm removing this shieeeet soon :shipit:

File: app/models/organisation.rb [hunk 9]
```diff
@@ -2,17 +2,19 @@ class Organisation < ActiveRecord::Base
   mount_uploader :picture, ThumbnailUploader
   has_many :events
   has_and_belongs_to_many :users
+  resourcify
```
Comment (JensDebergh) [inline app/models/organisation.rb]: What does this do?

File: app/models/role.rb [hunk 10]
```diff
@@ -0,0 +1,10 @@
+class Role < ActiveRecord::Base
+  has_and_belongs_to_many :users, :join_table => :users_roles
```
Comment (JensDebergh) [inline app/models/role.rb]: The reason why this is prefered is we can add additional fields later on without many database changes.

File: app/models/role.rb [hunk 11]
```diff
@@ -0,0 +1,10 @@
...
+  belongs_to :resource, :polymorphic => true
```
Comment (JensDebergh) [inline app/models/role.rb]: :police_car:

File: app/models/role.rb [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
+
+  validates :resource_type,
+            :inclusion => { :in => Rolify.resource_types },
+            :allow_nil => true
```
Comment (JensDebergh) [inline app/models/role.rb]: Please do so.

File: app/models/ticket.rb [hunk 13]
```diff
@@ -56,6 +56,7 @@ def self.ticket_price
   def self.add_to_revenue
     Revenue.create(
       event_id: @event.id,
+      name: @ticket.name,
```
Comment (JensDebergh) [inline app/models/ticket.rb]: self.name

**Comment by JensDebergh** [comment]
This probably works, but we need to cleanup heavily. 
This looks like a donkey but I want a unicorn. :baby:

File: app/models/ticket.rb (hunk 13 above)
Comment (FalkoJoseph) [inline app/models/ticket.rb]: This ain't gonna work: `undefined method "name"`

File: config/initializers/rolify.rb [hunk 14]
```diff
@@ -0,0 +1,7 @@
+Rolify.configure do |config|
```
Comment (FalkoJoseph) [inline config/initializers/rolify.rb]: @JensDebergh I don't understand. How can I fix this?

File: app/helpers/application_helper.rb (hunk 7 above)
Comment (FalkoJoseph) [inline app/helpers/application_helper.rb]: The thing is, the Rails flash object keys aren't compatible with Bootstrap 3 and above. They use `error` for `danger`, `notice` for `info` and `alert` for `warning`.

File: app/controllers/users_controller.rb [hunk 15]
```diff
@@ -1,7 +1,61 @@
...
+  end
 
   def show
+  end
+
+  def new
+    @user = User.new
+  end
+
+  def create
```
Comment (FalkoJoseph) [inline app/controllers/users_controller.rb]: I'll refactor this later.

File: app/controllers/events_controller.rb (hunk 2 above)
Comment (FalkoJoseph) [inline app/controllers/events_controller.rb]: Can you provide an example please?

File: app/controllers/organisations_controller.rb (hunk 3 above)
Comment (FalkoJoseph) [inline app/controllers/organisations_controller.rb]: The linter says: `Indent the first parameter one step more than the start of the previous line.` both Hound and my local Rubocop seem to use this convention.

File: app/controllers/registrations_controller.rb (hunk 4 above)
Comment (FalkoJoseph) [inline app/controllers/registrations_controller.rb]: We'll refactor this in another branch.

File: app/helpers/application_helper.rb (hunk 8 above)
Comment (FalkoJoseph) [inline app/helpers/application_helper.rb]: :+1: go ahead!

File: app/models/organisation.rb (hunk 9 above)
Comment (FalkoJoseph) [inline app/models/organisation.rb]: In the resource models you want to apply roles on you have to place `resourcify`.

https://github.com/RolifyCommunity/rolify#32-configure-your-resource-models

File: config/initializers/rolify.rb (hunk 14 above)
Comment (JensDebergh) [inline config/initializers/rolify.rb]: Hound is just warning you that you are using a variable that isn't being used.

Mainly the config variable is never used in the code block. So you could remove the entire initializer file cause you aren't doing anything to configure rolify besides the default.

File: app/helpers/application_helper.rb (hunk 7 above)
Comment (JensDebergh) [inline app/helpers/application_helper.rb]: The flash object is just a beefed up hash object so you can put in whatever you want.

If you say flash[:danger] = "Something went wrong"
Your iteration over the flash object will contain the key "danger" and the value "Something went wrong".

Just put in whatever you want!

File: app/controllers/events_controller.rb (hunk 2 above)
Comment (JensDebergh) [inline app/controllers/events_controller.rb]: create migration

def change
  change_column :events, :status, default: "open", null: false
end

File: app/models/ticket.rb (hunk 13 above)
Comment (JensDebergh) [inline app/models/ticket.rb]: Didn't see the class method there sorry :shipit:

**Pull request merged**